Framework/Kernel/test/EnumeratedStringTest.h

```diff
@@ -188,17 +188,21 @@ class EnumeratedStringTest : public CxxTest::TestSuite {
     TS_ASSERT(!functionOfString("Bundt Cake"));
   }
 
-  // void testEnumCount() {
-  //   // windows is a fake operating system that doesn't know how to compile this test
-  //   // this is a perfectly cromulent test
-  //   // try removing enum_count -- should give a compiler error
-  //   enum class Letters { a, b, enum_count };
-  //   static std::string letters[2] = {"a", "b"};
-  //   enum Graphia { alpha, beta, enum_count };
-  //   static std::string graphia[2] = {"alpha", "beta"};
-  //   EnumeratedString<Letters, letters> l1("a");
-  //   EnumeratedString<Graphia, graphia> l2("alpha");
-  //   TS_ASSERT_DIFFERS(l1, l2);
+  void testEnumCount() {
+    // windows doesn't know how to compile this test
+#ifndef _WIN32
+    // try removing enum_count -- should give a compiler error
+    enum class Letters{a, b, enum_count};
+    static std::string letters[2] = {"a", "b"};
+    enum Graphia{alpha, beta, enum_count};
+    static std::string graphia[2] = {"alpha", "beta"};
```

Comment:
Having looked at this on windows, the compiler error given in this test is due to the static c-style arrays of `std::string`, `letters` and `graphia` being defined inside the function. This is the only instance of this being done in these tests.

Given that declaring a static variable outside of the class/function makes it available to the whole translation unit, I guess this makes it work with the way you are passing the c-style arrays as template parameters. There must be an inconsistency between compilers with how static variables declared inside a function are handled in relation to template parameters.

I haven't ever seen c-style arrays of this kind passed as template parameters. As far as I'm aware, in c++ 17 the standard only specifies type parameters, integral values and pointers as valid template parameters. I suspect this is undefined behaviour, hence the inconsistency.

Is there another better defined way that the `EnumeratedString` class could be implemented, It's probably best we try to avoid c-style arrays in new code added?

Reply:
If you approve, I think I would prefer to use `std::array` instead of `std::vector`, and the reason is because it generates compiler errors, instead of runtime errors, when the enum and array have different lengths.  For instance:
``` cpp
template <class E, const std::vector<std::string>* names> class EnumeratedString{
public:
  EnumeratedString() {
    assert(size_t(E::enum_count) == names->size());
  }
}

void failIfWrongNumbers() {
  enum class Letters{a, b, c, enum_count};
  static const std::vector<std::string> letters {"a", "b"};
  typedef EnumeratedString<Letters, &letters> LETTERS;
  TS_ASSERT_THROWS_NOTHING(LETTERS l1);
}
```
not only compiles, but runs and passes.  Whereas
``` cpp
template <class E, const std::array<std::string, size_t(E::enum_count)>* names> class EnumeratedString{
public:
  EnumeratedString() {}
}

void failIfWrongNumbers() {
  enum class Letters{a, b, c, enum_count};
  static const std::array<std::string, size_t(Letters::enum_count)> letters {"a", "b"};
  typedef EnumeratedString<Letters, &letters> LETTERS;
  TS_ASSERT_THROWS_NOTHING(LETTERS l1);
}
```
will not even compile.

Reply:
However, the advantage of using a vector is that it serves double use for the `StringListValidator`